From 251cf60c57881bb03a4946e38b898ec1025a72f6 Mon Sep 17 00:00:00 2001
From: John Tyler <jetyler@ualberta.ca>
Date: Thu, 28 May 2026 10:08:22 -0600
Subject: [PATCH 1/2] hw/nvme: add properties to NvmeCtrl for p2p to device
 vram

allows the user to specify vram-dev=<id of pci device>,vram-bus=<bus #
of pcie device> when adding an emulated NVMe in QEMU to initiate p2p on
GPU VRAM
---
 hw/nvme/ctrl.c | 3 +++
 hw/nvme/nvme.h | 8 ++++++++
 2 files changed, 11 insertions(+)

diff --git a/hw/nvme/ctrl.c b/hw/nvme/ctrl.c
index ef0f938f03..c2c8469184 100644
--- a/hw/nvme/ctrl.c
+++ b/hw/nvme/ctrl.c
@@ -9530,6 +9530,9 @@ static void nvme_exit(PCIDevice *pci_dev)
 
 static const Property nvme_props[] = {
     DEFINE_BLOCK_PROPERTIES(NvmeCtrl, namespace.blkconf),
+    DEFINE_PROP_LINK("vram-dev", NvmeCtrl, vram_dev, TYPE_PCI_DEVICE,
+                     PCIDevice *),
+    DEFINE_PROP_UINT8("vram-bar", NvmeCtrl, vram_bar, 0),
     DEFINE_PROP_LINK("pmrdev", NvmeCtrl, pmr.dev, TYPE_MEMORY_BACKEND,
                      HostMemoryBackend *),
     DEFINE_PROP_LINK("subsys", NvmeCtrl, subsys, TYPE_NVME_SUBSYS,
diff --git a/hw/nvme/nvme.h b/hw/nvme/nvme.h
index 1d1e38eb90..f767aa3f96 100644
--- a/hw/nvme/nvme.h
+++ b/hw/nvme/nvme.h
@@ -624,6 +624,14 @@ typedef struct NvmeCtrl {
         hwaddr            cba;
     } pmr;
 
+    /*
+     * Optional VFIO passthrough device whose BAR is the source/sink for
+     * P2P DMA (e.g. a GPU's VRAM aperture). Resolved from the "vram-dev"
+     * link property; "vram-bar" selects which BAR.
+     */
+    PCIDevice *vram_dev;
+    uint8_t    vram_bar;
+
     uint8_t     aer_mask;
     NvmeRequest **aer_reqs;
     QTAILQ_HEAD(, NvmeAsyncEvent) aer_queue;
-- 
2.43.0

